File: Data Structures & Algorithms/generate-parentheses/submission-0.py
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        ret = []
        
        def backtrack(i, j, result):
            if i == n and j == n:
                ret.append(''.join(result))
                return
            if i<n:
                result.append('(')
                backtrack(i+1, j, result)
                result.pop()
            if j<i:
                result.append(')')
                backtrack(i, j+1, result)
                result.pop()

        backtrack(0, 0, [])
        return ret

File: Data Structures & Algorithms/combinations-of-a-phone-number/submission-0.cpp
class Solution {
private:
    map<char, string> digit_to_char = {
        {'2', "abc"},
        {'3', "def"},
        {'4', "ghi"},
        {'5', "jkl"},
        {'6', "mno"},
        {'7', "pqrs"},
        {'8', "tuv"},
        {'9', "wxyz"}
    };
    void backtrack(string& digits, int i, string curr, vector<string>& ret) {
        if ( i == digits.size() ) {
            ret.push_back(curr);
            return;
        }
        for (char c : digit_to_char[digits[i]]) {
            backtrack(digits, i+1, curr+c, ret);
        }   
    }
public:
    vector<string> letterCombinations(string digits) {
        vector<string> ret;
        if (digits=="") return ret;
        backtrack(digits, 0, "", ret);
        return ret;
    }
};
